app/services/ai_backend/anthropic.rb

```diff
@@ -13,9 +13,14 @@ def self.client
   end
 
   def initialize(user, assistant, conversation, message)
-    raise ::Anthropic::ConfigurationError if user.anthropic_key.blank?
     begin
-      @client = self.class.client.new(access_token: user.anthropic_key)
+      if assistant.api_service.blank?
```

Comment:
This branching doesn't feel quite right. Why are there any assistants that do not have an `api_service`? Why don't we just create api_services for all of our system assistants? If we keep this inconsistency then we'll have to add these sprinkle these conditional checks throughout the code, whereas if we resolve the inconsistency then we can always assume an assistant has an api_service

Reply:
Not sure, it seems to me then we'll have the same thing with the OpenAI/Anthropic URL; we don't want to configure those URLs in our database do we?

Reply:
I think we should configure those URLs in our database. The fact that we have two different ways of getting a URL is a good clue. If some assistant’s don’t have an api_service and others do, we’ll surely end up adding more conditionals over time to account for that difference.

If someone wanted to switch from OpenAI’s servers to using Microsoft Azure for all their GPT-4 needs, this would be the place to change it. It would be really nice to be able to edit that within the app. Also, I think we’ll inevitably end up adding additional logic to the API Service abstraction. Now that it exists, it’s a good place to house certain things. For example, one thing I’m working on is voice mode for the app. I did some digging and discovered that ChatGPT layers on some additional custom instructions when you activate the microphone. The GPT is told “the text you generate will be spoken so don’t use bullet points”.

This is not something I want to do per assistant or even per language model, I think it’s a setting which  I would add to the api_service model. I’ll pre-populate some “voice instructions” will get concatenated to your Assistant’s custom instructions, but if you want to tweak that setting it would get tweaked at this low level and apply to all assistants.

Reply:
If you're thinking about changing these URLs then they would need to be user-level. So every user will have these user-level api-services, implying the existing language_models records we created with the db-migration will be moved(copied) to user-level too then?  It does sound consistent, but I wanted to make sure I understand your suggestion right.

Reply:
Yes, that's right. So the Assistant > LanguageModel > APIService stack is all per-user. (Maybe some future "teams" features could cause us to revisit that with a more sophisticated permissions model...)

Reply:
Yes, that's right. So the Assistant > LanguageModel > APIService stack is all per-user. (Maybe some future "teams" features could cause us to revisit that with a more sophisticated permissions model...)